core/modules/layout/layout.admin.inc

```diff
@@ -1445,6 +1453,40 @@ function layout_block_add_page(Layout $layout, $renderer_name, $region_name, $bl
   return $output;
 }
 
+/**
+ * Helper to add prefixes to the block titles shown in the "Add block" dialog.
+ *
+ * Note: the "Field:" prefix for fields is already prepended in
+ * field_get_block_list().
+ *
+ * @see layout_block_add_page()
+ * @see field_get_block_list()
+ */
+function _layout_block_add_block_title_prefix($block_module, $block_info) {
+  if ($block_module == 'views') {
+    // Check if a "View:" prefix has already been added to the block
+    // description (see views_plugin_display_block::execute_hook_block_list)
+    // and add one if not.
+    $view_block_expected_prefix = "View: ";
+    $view_block_prefix_exists = strncmp($block_info['info'], $view_block_expected_prefix, strlen($view_block_expected_prefix)) === 0;
+    if (!$view_block_prefix_exists) {
+      $block_info['info'] = t('View: @block_title', array('@block_title' => $block_info['info']));
+    }
+  }
+  // Custom menus are coming from menu_block_info(), while system-defined menus
+  // from system_block_info().
+  if ($block_module == 'menu' || ($block_module == 'system' && isset($block_info['type']) && $block_info['type'] == 'menu')) {
```

Comment:
I'm not sure I like this approach, it's a bit messy. Can we leave this as simply `$block_module == 'menu'` and then add the same prefix in system module for the menu blocks added?